CONSOLE-3383: Minimal Dynamic Plugin for the CronTab CRD

raspbeep opened the pull request:
Adds minimal dynamic plugin for CronTab CRD.

https://issues.redhat.com/browse/CONSOLE-3383

Signed-off-by: Pavel Kratochvil <pakratoc@redhat.com>

jcaianirh reviewed (changes requested):
@raspbeep I looked through the readme to install and there are refs to helm installs. I think the install section was taken from another plugin and needs to be cleaned up.

spadgett commented on README.md:
````diff
@@ -0,0 +1,96 @@
+the plugin to a cluster by instantiating the template:
+
+```sh
+oc process -f oc-manifest.yaml \
````
This looks wrong. oc-manifest.yaml is not a template, so you shouldn't use `oc process`.

Is there a reason not to use the helm chart from the template repository?

spadgett commented on README.md:
```diff
@@ -0,0 +1,96 @@
+
+1. `yarn test-cypress-docker`
+2. Navigate to <http://localhost:10000>
+3. login with password `kubevirt` (no need for username)
```
Why is kubevirt the password?

spadgett commented on README.md:
```diff
@@ -0,0 +1,96 @@
+
+#### Cypress Testing inside the container
+
+1. `yarn test-cypress-docker`
```
I don't see this command in package.json

spadgett commented on README.md:
````diff
@@ -0,0 +1,96 @@
+1. Create a `dev.env` file inside the `.devcontainer` folder with the correct values for your cluster:
+
+```bash
+OC_PLUGIN_NAME=kubevirt-plugin
````
Why kubevirt?

spadgett commented on README.md:
````diff
@@ -0,0 +1,96 @@
+1. Build the image:
+
+   ```sh
+   docker build -t quay.io/crontab-plugin/crontab-plugin:latest .
````
Your quay repository won't be crontab-plugin. It's probably your user ID. We should leave this as a variable in the example.

spadgett commented on README.md:
````diff
@@ -0,0 +1,96 @@
+  | oc create -f -
+```
+
+Once deployed, patch the
+[Console operator](https://github.com/openshift/console-operator)
+config to enable the plugin.
````
If we use the helm chart, this isn't necessary.

spadgett commented on package.json:
```diff
@@ -0,0 +1,87 @@
+  "consolePlugin": {
+    "name": "crontab-plugin",
+    "version": "0.0.1",
+    "displayName": "OpenShift CronTab Plugin",
+    "description": "CronTab plugin for OpenShift Console",
```
Let's improve the description here to make it clear that it's a sample plugin for a custom resource and describe what capabilities it adds.

spadgett commented on package.json:
```diff
@@ -0,0 +1,87 @@
+      "useCronTabActionsProvider": "./views/hooks/useCronTabActionsProvider.tsx"
+    },
+    "dependencies": {
+      "@console/pluginAPI": "*"
```
I don't believe it's correct to use `*` anymore, cc @vojtechszocs

raspbeep replied:
@vojtechszocs would this be fine?
```
  "dependencies": {
    "@console/pluginAPI": "^4.11"
  }
```

spadgett commented on src/models/CronTabModel.ts:
```diff
@@ -0,0 +1,25 @@
+  apiVersion: 'v1',
+  apiGroup: 'stable.example.com',
+  plural: 'crontabs',
+  abbr: 'crontabs',
```
`abbr` should be short, like `CT` I think

spadgett commented on src/models/CronTabModel.ts:
```diff
@@ -0,0 +1,25 @@
+  crd: true,
+};
+
+export type CronTabKind = K8sResourceCommon & {
```
Let's create a better type definition for the actual shape of the resource.

spadgett reviewed (commented):
Thanks @raspbeep. I haven't looked closely at the code yet, but a few comments on the README and package.json.

spadgett commented on src/utils/Loading.tsx:
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+
+const Loading: React.FC = () => (
```
I wouldn't copy in the loader from console, and it's a bad practice for plugins to reference console styles directly since they might change.

We should look at switching console to use the PatternFly loader, and I'd use that here in the plugin (even if different from console today).

spadgett commented on src/utils/components/ModalProvider/ModalProvider.tsx:
```diff
@@ -0,0 +1,76 @@
+import * as React from 'react';
```
Was this file copied over from console?

spadgett commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,10 @@
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+const modelToRef = (obj: K8sModel) => `${obj.apiGroup}~${obj.apiVersion}~${obj.kind}`;
```
Where do we need to use references? We've been trying to avoid exposing that in the SDK, so we should look at what gap we have.

raspbeep replied:
It is used for `ListPageCreate` component from `@dynamic-plugin-sdk` which requires `GroupVersionKind` string (in our case `stable.example.com~v1~CronTab`)

spadgett replied:
@raspbeep Please open a bug for this component. We shouldn't be exposing `GroupVersionKind` strings in the API, so we'll want to deprecate this property and replace it with a prop where these values are specified separately.

spadgett commented on src/utils/hooks/useCronTabTranslation.ts:
```diff
@@ -0,0 +1,13 @@
+ * a function to perform translation to 'plugin__crontab-plugin' namespace
+ * @param value string to translate
+ */
+// skipcq: JS-C1002
```
What is `skipcq`?

spadgett commented on src/views/CronTabDetails.tsx:
```diff
@@ -0,0 +1,65 @@
+        component: CronTabYAMLPage,
+      },
+    ],
+    [],
```
I think we need `t` as a dependency here to force re-render when the language changes?

Since others are using this as a model, we should check what's right here.

spadgett commented on src/views/components/CronTabActions/CronTabActions.scss:
```diff
@@ -0,0 +1,11 @@
+.console-cron-tab-actions {
```
We should avoid overriding PF styles. Why is this needed?

spadgett commented on src/views/CronTabList/CronTabList.scss:
```diff
@@ -0,0 +1,16 @@
+.co-operator-details__actions {
```
We should avoid overriding console or PF styles. Why is this needed?

spadgett commented on src/utils/constants.ts:
```diff
@@ -0,0 +1 @@
+export const DEFAULT_NAMESPACE = 'default';
```
How is this used?

We should avoid putting anything in the `default` namespace

spadgett commented on src/views/modals/LabelsModal/LabelsModal.tsx:
```diff
@@ -0,0 +1,158 @@
+import * as React from 'react';
```
Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the labels modal.

spadgett commented on src/views/modals/DeleteModal/DeleteModal.tsx:
```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
```
Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the delete modal.

spadgett commented on src/views/modals/AnnotationsModal/AnnotationsModal.tsx:
```diff
@@ -0,0 +1,110 @@
+import * as React from 'react';
```
Was this copied over from console? We should avoid reimplementing console functionality and instead open an issue to expose the annotations modal.

spadgett reviewed (changes requested):
I'm still working through the review, but wanted to get you feedback. We should try to keep the plugin simple and avoid copying content over from openshift/console. If there are missing capabilities, we need to open issues to track. Ideally we should be using PF components and console SDK components with minimal code in the plugin itself.

raspbeep commented on console-extensions.json:
```diff
@@ -0,0 +1,87 @@
+    }
+  },
+  {
+    "type": "console.navigation/resource-ns",
```
What the `type` value should be then?

raspbeep commented:
@spadgett I removed the DetailsPage and replaced it with pure PF components. I also edited the readme and the description in package.json. The ModalProvider and various Modals(Labels, Annotations and Delete) stay for now and will be replaced once they are properly exposed in the `dynamic-plugin-sdk`.

spadgett commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM docker.io/library/node:16 AS build
```
We should rebase on the current plugin template, which uses different images.

spadgett commented on README.md:
```diff
@@ -0,0 +1,79 @@
+## Deployment on cluster
+
+
+### Option 1: Installing the Helm Chart (WIP - does not work yet)
```
Do we know why it doesn't work? I think we want to solve this before merging.

spadgett commented on README.md:
````diff
@@ -0,0 +1,79 @@
+You can deploy the plugin to a cluster by applying `oc-manifest.yaml`.
+
+```sh
+oc apply -f oc-manifest.yaml
````
nitpick: I'd probably rename to `manifest.yaml`.

spadgett commented on README.md:
```diff
@@ -0,0 +1,79 @@
+
+## Docker image
+
+Before you can deploy your plugin on a cluster, you must build an image and
+push it to an image registry.
```
This statement more or less contradicts the instructions at the top of the README.

spadgett commented on src/utils/hooks/useCronTabTranslation.ts:
Again, let's not use utils and put this is a folder like i18n or similar.

spadgett commented on src/views/CronTabDetails/CronTabPageTitle.tsx:
```diff
@@ -0,0 +1,46 @@
+      <div className="co-m-nav-title co-m-nav-title--detail co-m-nav-title--breadcrumbs">
+        <span className="co-m-pane__heading">
+          <h1 className="co-m-pane__name co-resource-item">
+            <span className="co-m-resource-icon co-m-resource-icon--lg">{'DS'}</span>
```
This is a gap in our SDK if we don't provide a way to render a resource title like other pages. We need components to render this rather than hard-coding internal console styles.

spadgett commented on src/views/CronTabDetails/CronTabPageTitle.tsx:
```diff
@@ -0,0 +1,46 @@
+          <BreadcrumbItem>{t('CronTab Details')}</BreadcrumbItem>
+        </Breadcrumb>
+      </div>
+      <div className="co-m-nav-title co-m-nav-title--detail co-m-nav-title--breadcrumbs">
```
We should avoid any `co-` prefixed classes since these are internal and not public APIs. They can change and break plugins from version to version. If we need to wait for the SDK to expose details page components, we should probably wait, or just use PF components for now.

spadgett commented on src/views/CronTabDetailsPage.tsx:
```diff
@@ -0,0 +1,116 @@
+  const { t } = useCronTabTranslation();
+  const { createModal } = useModal();
+
+  const onEditLabels = () => {
```
We should remove this from the PR and wait until it's available in the SDK.

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+    namespaced: true,
+    namespace,
+  });
+  // const { t } = useTranslation();
```
Need to update this file to use translations

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+        id: 'name',
+        transforms: [sortable],
+        sort: 'metadata.name',
+        props: { className: 'pf-m-width-15' },
```
Do we need these widths? By default, won't all columns get the same width?

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+  const columns: TableColumn<K8sResourceCommon>[] = React.useMemo(
+    () => [
+      {
+        title: 'Name',
```
Needs translation

spadgett commented on src/views/utils/MetadataLabels/MetadataLabels.tsx:
```diff
@@ -0,0 +1,25 @@
+        </LabelGroup>
+      ) :
+      (
+        <span className="metadata-labels-no-labels">No labels</span>
```
Can we use a PF style instead of a custom style?

spadgett commented on src/views/modals/TabModal/TabModal.tsx:
```diff
@@ -0,0 +1,140 @@
+    submitBtnVariant,
+    titleIconVariant,
+  }) => {
+    const [isSubmitting, setIsSubmitting] = React.useState(false);
```
nitpick:
```suggestion
    const [isSubmitting, setSubmitting] = React.useState(false);
```

spadgett commented on src/views/modals/TabModal/TabModal.tsx:
```diff
@@ -0,0 +1,140 @@
+
+      const promise = onClose();
+
+      if (promise) promise?.catch(setError);
```
```suggestion
      promise?.catch(setError);
```

spadgett commented on src/views/modals/TabModal/TabModal.tsx:
```diff
@@ -0,0 +1,140 @@
+                </Alert>
+              </StackItem>
+            )}
+            <StackItem>
```
Do we not let you update any of the fields in this modal?

spadgett commented on src/views/modals/LabelsModal/utils.ts:
Let's remove any custom logic for editing labels in this PR.

spadgett commented on src/views/CronTabList/CronTabRowActions.tsx:
Let's wait until we have SDK support before adding all of the actions.

spadgett reviewed (changes requested):
Let's break this up into smaller PRs (list page, details page, actions, etc.) and only merge things when ready. If there are missing things in the SDK, it's better to hold the corresponding PR. This is a sample others will base their plugins on, so we shouldn't merge short-term workarounds for SDK gaps. They'll make their way into other plugin code that might not get updated.

spadgett commented:
These changes were rolled into #2. Closing...
